### PR TITLE
[SPARK-32106][SQL] Implement script transform in sql/core

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

 * Implement `SparkScriptTransformationExec` based on `BaseScriptTransformationExec`
 * Implement `SparkScriptTransformationWriterThread` based on `BaseScriptTransformationWriterThread` of writing data
 * Add rule `SparkScripts` to support convert script LogicalPlan to SparkPlan in Spark SQL (without hive mode)
 * Add `SparkScriptTransformationSuite` test spark spec case 
 * add test in `SQLQueryTestSuite`

And we will close #29085 .

### Why are the changes needed?
Support user use Script Transform without Hive

### Does this PR introduce _any_ user-facing change?
User can use Script Transformation without hive in no serde mode.
Such as : 
**default no serde **
```
SELECT TRANSFORM(a, b, c)
USING 'cat' AS (a int, b string, c long)
FROM testData
```
**no serde with spec ROW FORMAT DELIMITED**
```
SELECT TRANSFORM(a, b, c)
ROW FORMAT DELIMITED
FIELDS TERMINATED BY '\t'
COLLECTION ITEMS TERMINATED BY '\u0002'
MAP KEYS TERMINATED BY '\u0003'
LINES TERMINATED BY '\n'
NULL DEFINED AS 'null'
USING 'cat' AS (a, b, c)
ROW FORMAT DELIMITED
FIELDS TERMINATED BY '\t'
COLLECTION ITEMS TERMINATED BY '\u0004'
MAP KEYS TERMINATED BY '\u0005'
LINES TERMINATED BY '\n'
NULL DEFINED AS 'NULL'
FROM testData
```

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
FYI @maropu  @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1105]: Could you add some indents for readability?
```
SELECT TRANSFORM(a, b, c)
  ROW FORMAT DELIMITED
  FIELDS TERMINATED BY '\t'
  COLLECTION ITEMS TERMINATED BY '\u0002'
  MAP KEYS TERMINATED BY '\u0003'
  LINES TERMINATED BY '\n'
  NULL DEFINED AS 'null'
  USING 'cat' AS (a, b, c)
  ROW FORMAT DELIMITED
  FIELDS TERMINATED BY '\t'
  COLLECTION ITEMS TERMINATED BY '\u0004'
  MAP KEYS TERMINATED BY '\u0005'
  LINES TERMINATED BY '\n'
  NULL DEFINED AS 'NULL'
FROM testData
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1143]: ditto:
```
SELECT TRANSFORM(a, b, c)
  ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde'
  WITH SERDEPROPERTIES(
    "separatorChar" = "\t",
    "quoteChar" = "'",
    "escapeChar" = "\\")
  USING 'cat' AS (a, b, c)
  ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde'
  WITH SERDEPROPERTIES(
    "separatorChar" = "\t",
    "quoteChar" = "'",
    "escapeChar" = "\\")
FROM testData
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:42]: Could you add this param in `@param` above?

File: sql/core/src/test/resources/sql-tests/inputs/transform.sql
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/transform.sql:24]: nit: two indents in the other xxx.sql files?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:42]: > Could you add this param in `@param` above?

How about just add this in BasicScripTransformationExec?
Since HiveScriptTransformationExec's param is not correct too

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1105]: > SELECT TRANSFORM(a, b, c)
>   ROW FORMAT DELIMITED
>   FIELDS TERMINATED BY '\t'
>   COLLECTION ITEMS TERMINATED BY '\u0002'
>   MAP KEYS TERMINATED BY '\u0003'
>   LINES TERMINATED BY '\n'
>   NULL DEFINED AS 'null'
>   USING 'cat' AS (a, b, c)
>   ROW FORMAT DELIMITED
>   FIELDS TERMINATED BY '\t'
>   COLLECTION ITEMS TERMINATED BY '\u0004'
>   MAP KEYS TERMINATED BY '\u0005'
>   LINES TERMINATED BY '\n'
>   NULL DEFINED AS 'NULL'
> FROM testData

How about make `USING` clause without indents?
```
SELECT TRANSFORM(a, b, c)
  ROW FORMAT DELIMITED
  FIELDS TERMINATED BY '\t'
  COLLECTION ITEMS TERMINATED BY '\u0002'
  MAP KEYS TERMINATED BY '\u0003'
  LINES TERMINATED BY '\n'
  NULL DEFINED AS 'null'
USING 'cat' AS (a, b, c)
  ROW FORMAT DELIMITED
  FIELDS TERMINATED BY '\t'
  COLLECTION ITEMS TERMINATED BY '\u0004'
  MAP KEYS TERMINATED BY '\u0005'
  LINES TERMINATED BY '\n'
  NULL DEFINED AS 'NULL'
FROM testData
```

File: sql/core/src/test/resources/sql-tests/inputs/transform.sql
Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/transform.sql:24]: > nit: two indents in the other xxx.sql files?

change to 2 indents

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1143]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:42]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:36]: nit: `* @param child logical plan whose output is transformed`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala:48]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:37]: nit: `The` -> `the`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala:36]: > nit: `* @param child logical plan whose output is transformed`

Done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala:48]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala]: Done

File: sql/core/src/test/resources/sql-tests/results/transform.sql.out
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/results/transform.sql.out:195]: Can `ROW FORMAT DELIMETED` with output schema work correctly?

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:195]: Also, could you add test cases for the parser, too? https://github.com/apache/spark/pull/29414/files#diff-36e2b29ae675caaa1fce16e74fbd8710R1135

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:195]: @maropu

see two more bug founded
 https://issues.apache.org/jira/browse/SPARK-32607
https://issues.apache.org/jira/browse/SPARK-32608

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:195]: Oh, I see...

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:195]: > Also, could you add test cases for the parser, too? https://github.com/apache/spark/pull/29414/files#diff-36e2b29ae675caaa1fce16e74fbd8710R1135

Add some UT in transform.sql

**Comment by maropu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
Any update?

**Comment by AngersZhuuuu** [comment]
ping @maropu  @HyukjinKwon  @cloud-fan

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
ping  @maropu @cloud-fan  Is ok to merge this pr and  https://github.com/apache/spark/pull/29421 
We can start the rest  work。

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @wangyum too

**Comment by AngersZhuuuu** [comment]
ping @maropu @cloud-fan Is ok for this pr to merge?  @alfozan will raise pr about spark native serde after this pr merged.

**Review by maropu — APPROVED** [review]
I've check this again and I don't have no more comment.

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @wangyum

**Comment by HyukjinKwon** [comment]
I haven't checked it very closely but looks good to go. WDYT @wangyum?

**Comment by maropu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @dongjoon-hyun

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
also ping @HyukjinKwon

**Comment by AngersZhuuuu** [comment]
@maropu @HyukjinKwon @cloud-fan Have resolve the conflicts, can we start these work again?

**Comment by maropu** [comment]
Looks fine to me. Btw, if we merge this, we can close #29085? If so, please add "Closes #29085" in the PR description.

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
> Looks fine to me. Btw, if we merge this, we can close #29085? If so, please add "Closes #29085" in the PR description.

Yes, we can close #29085  after merge this, and for some other small point under https://issues.apache.org/jira/browse/SPARK-31936 . I will raise separated PR based on this.

**Comment by AngersZhuuuu** [comment]
FYI @maropu This [build](https://github.com/apache/spark/pull/29414#issuecomment-726664932) failed  because  https://github.com/apache/spark/pull/29421 was merged and  have updated `transform.sql.output`

**Comment by AngersZhuuuu** [comment]
Any update?

**Comment by AngersZhuuuu** [comment]
Can we restart this part?

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I am good with this change. I don't mind if somebody merges.

**Comment by AngersZhuuuu** [comment]
Hope merge this and start next work

**Comment by AngersZhuuuu** [comment]
gentle ping @maropu @cloud-fan

**Comment by alfozan** [comment]
gentle ping @tejasapatil @sameeragarwal

**Comment by maropu** [comment]
okay, I will merge this into master in a few days if nobody has more comments.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
okay, I believe we have much time to revisit this until the Spark v3.2.0 released, so I'll merge this into master for now so as to encourage @AngersZhuuuu 's following work. FYI:  @HyukjinKwon @cloud-fan @dongjoon-hyun @gatorsmile

**Comment by maropu** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Nice, I support your decision @maropu.

**Comment by dongjoon-hyun** [comment]
Thank you for informing, @maropu !

File: sql/core/src/test/resources/sql-tests/results/transform.sql.out
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: This seems to cause a test flakiness in GitHub Action. Could you take a look, @AngersZhuuuu and @maropu .
```
SQLQueryTestSuite.transform.sql
org.scalatest.exceptions.TestFailedException: transform.sql
Expected "...istent_command: comm[and not found]", but got "...istent_command: comm[]" Result did not match for query #2
SELECT TRANSFORM(a)
USING 'some_non_existent_command' AS (a)
FROM t
```

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: https://github.com/apache/spark/runs/1592987501

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: Ah, that's too bad. Thanks. I'll open a followup PR to fix it.

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: Thank you so much for taking a look at that, @maropu .

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: Please review this? https://github.com/apache/spark/pull/30896

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/transform.sql.out:37]: > Please review this: #30896

Thank you, @maropu 
Yesterday I also found some different between UT and `transform.sql`. `USING 'cat 1>&2'` works well in SQLQuerySuite but won't work in `transform.sql`

**Pull request closed without merging**
